### PR TITLE
Feature/provenance (no to be merged)

### PR DESCRIPTION
**Pull request by Kyrosh**

### CODE REVIEW

**Comment by splattater** [comment]
Hey @Kyrosh thank you for your pull request. Can you please

- create a module containg utils.py and all rdflib clones (evaluate.py ...)
- move git.py (replacing GitRepo) and graphs.py to core.py
- check if there is a solution for add/delete quads (via Flask API) that supports provenance without use of FileReference class. If so, delete FileReference from core.py
- delete deprecated.py
- are there other candidates that may be merged to a module? (serializer, parser?)

File: quit/quit.py [hunk 1]
```diff
@@ -45,7 +42,6 @@
 logger.addHandler(fh)
 logger.addHandler(ch)
 
-
```
Comment (white-gecko) [inline quit/quit.py]: Could you please also run the liter?

File: quit/quit.py [hunk 2]
```diff
@@ -406,7 +363,7 @@ def savedexit():
     Add methods you want to call on unexpected shutdown.
     """
     logger.info("Exiting store")
-    store.exit()
+    #store.exit()
```
Comment (white-gecko) [inline quit/quit.py]: Please don't add out-commented code.

File: quit/web/static/css/bootstrap-theme.min.css [hunk 3]
```diff
@@ -0,0 +1,6 @@
+/*!
```
Comment (white-gecko) [inline quit/web/static/css/bootstrap-theme.min.css]: Maybe we should not include a static version of bootstrap in the repo, but use something like this: https://pythonhosted.org/Flask-Bootstrap/

File: quit/web/modules/git.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    """Pull from remote.
+
+    Returns:
+        HTTP Response 201: If pull was possible
+        HTTP Response: 403: If pull did not work
+    """
+    #if current_app.config['repository'].pull():
+    #    return '', status.HTTP_201_CREATED
+    #else:
+    #    return '', status.HTTP_403_FORBIDDEN
```
Comment (white-gecko) [inline quit/web/modules/git.py]: Please don't add outcommented code

**Comment by Kyrosh** [comment]
@white-gecko It would be possible to provide the frontend as a seperate project. One possibility to do so: the quit store itself provides an API and a minimalistic server for the sparql endpoint. When the fronend is used it could provide its own server, only using the provided API and bypassing the quit store server (only one running). What do you think about this approach? Or we provide the frontend as a seperate package and check in the quit store if it is installed, if so, we use those endpoints instead. Both approaches would be sufficient, i think.

**Comment by white-gecko** [comment]
> @white-gecko It would be possible to provide the frontend as a seperate project. One possibility to do so:
> the quit store itself provides an API and a minimalistic server for the sparql endpoint. When the fronend is used it could provide its own server, only using the provided API and bypassing the quit store server (only one running). What do you think about this approach?

I thinks this sounds like the option I'm preferring. The Quit Store provides the necessary API and the front-end is using it. What do you mean by “bypassing the quit store server”?

> Or we provide the frontend as a seperate package and check in the quit store if it is installed, if so, we use those endpoints instead. Both approaches would be sufficient, i think.

What exactly do you mean by separate package?

Further I'm looking for the documentation of the newly introduced store modes. Could you please add them to the README in your branch?

**Comment by white-gecko** [comment]
> @white-gecko It would be possible to provide the frontend as a seperate project. One possibility to do so:
> the quit store itself provides an API and a minimalistic server for the sparql endpoint. When the fronend is used it could provide its own server, only using the provided API and bypassing the quit store server (only one running). What do you think about this approach?

I thinks this sounds like the option I'm preferring. The Quit Store provides the necessary API and the front-end is using it. What do you mean by “bypassing the quit store server”?

> Or we provide the frontend as a seperate package and check in the quit store if it is installed, if so, we use those endpoints instead. Both approaches would be sufficient, i think.

What exactly do you mean by separate package?

Further I'm looking for the documentation of the newly introduced store modes. Could you please add them to the README in your branch?

**Comment by Kyrosh** [comment]
> I thinks this sounds like the option I'm preferring. The Quit Store provides the necessary API and the front-end is using it. What do you mean by “bypassing the quit store server”?

By that I just meant that the frontend has its own flask server, which will be used instead of the one included in the quit store. So we only have  one server running at a time.

> What exactly do you mean by separate package?

It's basically the same idea as the previous one. Two seperate and indepentent projects and packages.
In the first scenario the frontend server is startet via its own quit.py (just as an example).
In the second one we could run the quit.py of the main project, which checks if the frontent package is installed and available. If so, it registers the frontend endpoints, or otherwise its own endpoints.
Therefore the user won't have to run a seperate project, but instead he/she only needs to install the frontend package (extension).

> Further I'm looking for the documentation of the newly introduced store modes. Could you please add them to the README in your branch?

Sure

File: quit/conf.py [hunk 5]
```diff
@@ -278,6 +288,32 @@ def getRepoPath(self):
         for s, p, o in self.sysconf.triples((None, property, None)):
             return str(o)
 
+    def getDefaultBranch(self):
+        """Get the path of Git repository from configuration.
+
+        Returns:
+            A string containig the path of the git repo.
+        """
```
Comment (white-gecko) [inline quit/conf.py]: I think this doc comment is not correct

File: quit/conf.py [hunk 6]
```diff
@@ -278,6 +288,32 @@ def getRepoPath(self):
...
+
+        for s, p, o in self.sysconf.triples((None, property, None)):
+            return str(o)
+
+    def getGlobalFile(self):
+        """Get the path of Git repository from configuration.
+
+        Returns:
+            A string containig the path of the git repo.
+        """
```
Comment (white-gecko) [inline quit/conf.py]: Same for this doc comment.

File: quit/conf.py [hunk 7]
```diff
@@ -164,7 +171,7 @@ def __readGraphIriFile(self, graphfile):
         Returns:
             graphuri: String with the graph URI
         """
-        if isfile(graphfile):
+        if isfile(graphfile):
```
Comment (white-gecko) [inline quit/conf.py]: This should be reported by the linter

File: quit/core.py [hunk 8]
```diff
@@ -317,6 +353,290 @@ def exit(self):
...
+        for name in self.repository.tags_or_branches:
+            initial_commit = self.repository.revision(name);            
+            commits = traverse(initial_commit, seen)
+
+            prov = self.changesets(commits)                          
+            self.store.addquads((s, p, o, c) for s, p, o, c in prov.quads())
+
+            #for commit in commits:
+                #(_, g) = commit.__prov__()
+                #self.store += g
```
Comment (white-gecko) [inline quit/core.py]: Could you please remove outcommented code

File: quit/git.py [hunk 9]
```diff
@@ -0,0 +1,638 @@
...
+            else:
+                pygit2.init_repository(path)
+
+        name = os.path.basename(path).lower()
+
+        self.name = name
+        self.path = path
+        self.params = params
+
+    def _callback(self, origin):
```
Comment (white-gecko) [inline quit/git.py]: I think we have a better solution for this already here: https://github.com/AKSW/QuitStore/blob/ff4598a4ca6dbe83619924df67d3d4f2816e5f58/quit/core.py#L654L697

File: quit/git.py [hunk 10]
```diff
@@ -0,0 +1,638 @@
...
+        # entities
+        for entity in self.node().entries(recursive=True):
+            for (entity_uri, entity_graph) in self.committer.__prov__():
+                g += entity_graph
+                g.add((entity_uri, PROV['wasGeneratedBy'], uri))
+
+        return (uri, g)
+
+
+class Signature(Base):
```
Comment (white-gecko) [inline quit/git.py]: Why is not the pygit2.Signature suitable here?

File: quit/quit.py [hunk 11]
```diff
@@ -205,66 +196,32 @@ def initialize(args):
                 logger.info('Git garbage collection could not be configured and was disabled')
                 logger.debug(e)
 
+    storemode = STORE_ALL
+    if args.disable_data_store:
+        storemode &= ~STORE_DATA
+    if args.disable_provenance_store:
+        storemode &= ~STORE_PROVENANCE
+
```
Comment (white-gecko) [inline quit/quit.py]: These modes should be documented in the README

File: quit/web/modules/endpoint.py [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+                res = graph.update(q)                          
+                quit.commit(graph, 'New Commit from QuitStore', branch_or_ref, ref, query=q)          
+                return '', 200
+            
+        except Exception as e:
+            current_app.logger.error(e)
+            current_app.logger.error(traceback.format_exc())
+            return "<pre>"+traceback.format_exc()+"</pre>", 400
+    else:
+        return render_template('sparql.html')
```
Comment (white-gecko) [inline quit/web/modules/endpoint.py]: This should move to new Quit UI

File: quit/web/modules/endpoint.py [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+                    return response   
+            else:
+                raise UnSupportedQueryType()
+            
+        except Exception as e:
+            current_app.logger.error(e)
+            current_app.logger.error(traceback.format_exc())
+            return "<pre>"+traceback.format_exc()+"</pre>", 400
+    else:
+        return render_template('provenance.html')
```
Comment (white-gecko) [inline quit/web/modules/endpoint.py]: This should move to new Quit UI

**Comment by white-gecko** [comment]
I've deleted some already resolved comments. We should continue the discussion in https://github.com/AKSW/QuitStore/pull/47

File: quit/web/static/css/bootstrap-theme.min.css (hunk 3 above)
Comment (white-gecko) [inline quit/web/static/css/bootstrap-theme.min.css]: #50

File: quit/quit.py (hunk 11 above)
Comment (white-gecko) [inline quit/quit.py]: #52

File: quit/web/modules/endpoint.py (hunk 12 above)
Comment (white-gecko) [inline quit/web/modules/endpoint.py]: #50

File: quit/web/modules/endpoint.py (hunk 13 above)
Comment (white-gecko) [inline quit/web/modules/endpoint.py]: #50

File: quit/quit.py (hunk 2 above)
Comment (white-gecko) [inline quit/quit.py]: #53

File: quit/web/modules/git.py (hunk 4 above)
Comment (white-gecko) [inline quit/web/modules/git.py]: #53

File: quit/core.py (hunk 8 above)
Comment (white-gecko) [inline quit/core.py]: #53

File: quit/git.py (hunk 9 above)
Comment (white-gecko) [inline quit/git.py]: #54

File: quit/conf.py (hunk 5 above)
Comment (white-gecko) [inline quit/conf.py]: #55

File: quit/conf.py (hunk 6 above)
Comment (white-gecko) [inline quit/conf.py]: #55

File: quit/git.py (hunk 10 above)
Comment (white-gecko) [inline quit/git.py]: #56

File: quit/core.py [hunk 14]
```diff
@@ -317,6 +353,290 @@ def exit(self):
...
+
+                        if self.config.checkStoremode(STORE_PROVENANCE | STORE_DATA):
+                            g.add((private_uri, PROV['specializationOf'], public_uri))
+                            g.add((private_uri, PROV['wasGeneratedBy'], commit_uri))
+                        if self.config.checkStoremode(STORE_DATA):
+                            g.addN((s, p, o, private_uri) for s, p, o in tmp.triples((None, None, None), context))
+                                                                       
+        return g
+   
+    def commit(self, graph, message, index, ref, **kwargs):
```
Comment (white-gecko) [inline quit/core.py]: #58

File: quit/core.py [hunk 15]
```diff
@@ -317,6 +353,290 @@ def exit(self):
...
+        #message = "\n".join(out)
+
+        author = self.repository._repository.default_signature
+        id = index.commit(message, author.name, author.email, ref=ref)
+
+        if id:            
+            self.repository._repository.set_head(id)            
+            if not self.repository.is_bare:                            
+                self.repository._repository.checkout(ref, strategy=pygit2.GIT_CHECKOUT_FORCE)
+            self.sync()
```
Comment (white-gecko) [inline quit/core.py]: #59

File: quit/git.py [hunk 16]
```diff
@@ -0,0 +1,638 @@
...
+        index = Index(self)
+        if revision:
+            index.set_revision(revision)
+        return index
+
+    def pull(self, remote_name='origin', branch='master'):
+        for remote in self._repository.remotes:
+            if remote.name == remote_name:
+                remote.fetch()
+                remote_master_id = self._repository.lookup_reference('refs/remotes/origin/%s' % (branch)).target
```
Comment (white-gecko) [inline quit/git.py]: #60

**Pull request merged**